Update customer source from hash

ErikPeterson opened the pull request:
Add the ability to use a hash to update a customer's default source, and spec to prove behavior

alexmamonchik commented:
Hi @ErikPeterson Thanks for your help. Can you check [live testing](https://github.com/rebelidealist/stripe-ruby-mock/wiki/Implementing-a-New-Behavior#live-testing) article and test your method in live mode. I did it and got error. 
```
1) StripeMock::Instance behaves like Customer API updates a stripe customer's card from a hash
     Failure/Error: original = Stripe::Customer.create(id: 'test_customer_update', source: gen_card_tk)
     Stripe::InvalidRequestError:
       Customer already exists.
     Shared Example Group: "Customer API" called from ./spec/support/stripe_examples.rb:14
     # ./spec/shared_stripe_examples/customer_examples.rb:301:in `block (2 levels) in <top (required)>'
```

For fix this error look at [strategies](https://github.com/rebelidealist/stripe-ruby-mock/tree/master/lib/stripe_mock/test_strategies) files

ErikPeterson commented:
@alexmamonchik Is there a particular test API key I should use for STRIPE_TEST_SECRET_KEY or will any Stripe test API key do?

alexmamonchik commented:
you can find keys [here](https://github.com/rebelidealist/stripe-ruby-mock/blob/master/.travis.yml). I use 
```shell
$  STRIPE_TEST_SECRET_KEY=sk_test_d8s39dj3fdf92j44hf9 rspec -t live
```
command for exec test. Comment some tests [here](https://github.com/rebelidealist/stripe-ruby-mock/blob/master/spec/support/stripe_examples.rb) before

ErikPeterson commented:
Hey Alex, I'm unable to reproduce the specific error you mentioned in live mode. Also, if I roll back my changes many of the Customer API examples fail in live mode, so I don't think that my patch is necessarily the root of the error you're seeing. Could you tell me which lines you commented out in `stripe_examples.rb`, your Ruby version, and the exact rspec command you're using to run the tests, so that I can see if the error you're getting is reproduceable?

alexmamonchik commented on spec/shared_stripe_examples/customer_examples.rb:
```diff
@@ -281,7 +281,7 @@ def gen_card_tk
   end
 
-  it "updates a stripe customer's card" do
+  it "updates a stripe customer's card from a token" do
```
I added `live: true` here

alexmamonchik commented on spec/shared_stripe_examples/customer_examples.rb:
```diff
@@ -297,6 +297,26 @@ def gen_card_tk
     expect(new_card.id).to_not eq(card.id)
   end
 
+  it "updates a stripe customer's card from a hash" do
```
and I added `live: true` here

alexmamonchik commented:
You have my command above. I commented all examples except https://github.com/SwedenUnltd/stripe-ruby-mock/blob/master/spec/support/stripe_examples.rb#L14

ErikPeterson commented:
Could you also give me your ruby version and your Gemfile.lock contents? With the exact same code and rspec command I'm getting different results, so there must be some difference in our environments that is causing the divergence in the errors.

alexmamonchik commented:
[Gemfile](https://gist.github.com/alexmamonchik/a7a17905c6a2705a93b6). `ruby 2.1.6p336 (2015-04-13 revision 50298) [x86_64-darwin14.0]`. really strange behaviour.

ErikPeterson commented:
Definitely weird. Looks like there's some possibly significant differences in our Gem versions (e.g. stripe). I'm going to match our versions and try to run down the issue tomorrow. Thanks for all the info!

ErikPeterson commented:
@alexmamonchik I've updated the spec and `StripeMock::TestStrategies::Base`. Stripe does not expect the `fingerprint` param to be sent when creating a source, which was part of the issue with the spec. I've also removed the explicit customer ids from the customer creation calls in the two Customer update specs, to avoid the `Customer already exists` error you were encountering.

alexmamonchik commented:
I have small updates in master branch. Please, update your PR.

ErikPeterson commented:
I've rebased off your master branch
